sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java

```diff
@@ -118,59 +111,34 @@ public void verifyUnmodified() {
-      // versa).
-      //
-      // To enable a reasonable comparison, we clone retainedObject again here, converting it to
-      // the same sort of T that the Coder<T> output when it created clonedObject.
-      T clonedPossiblyModifiedObject = CoderUtils.clone(coder, possiblyModifiedObject);
-
-      // If deepEquals() then we trust the equals implementation.
-      // This deliberately allows fields to escape this check.
-      if (Objects.deepEquals(clonedPossiblyModifiedObject, clonedOriginalObject)) {
-        return;
+      Object newStructuralValue = coder.structuralValue(possiblyModifiedObject);
+      if (originalStructuralValue.equals(newStructuralValue)) {
+          return;
       }
-
-      // If not deepEquals(), the class may just have a poor equals() implementation.
-      // So we next try checking their serialized forms. We re-serialize instead of checking
-      // encodedObject, because the Coder may treat it differently.
-      //
-      // For example, an unbounded Iterable will be encoded in an unbounded way, but decoded into an
-      // ArrayList, which will then be re-encoded in a bounded format. So we really do need to
-      // encode-decode-encode retainedObject.
-      if (Arrays.equals(
-          CoderUtils.encodeToByteArray(coder, clonedOriginalObject),
-          CoderUtils.encodeToByteArray(coder, clonedPossiblyModifiedObject))) {
-        return;
-      }
-
-      // If we got here, then they are not deepEquals() and do not have deepEquals() encodings.
-      // Even if there is some conceptual sense in which the objects are equivalent, it has not
-      // been adequately expressed in code.
-      illegalMutation(clonedOriginalObject, clonedPossiblyModifiedObject);
+      illegalMutation(originalStructuralValue, newStructuralValue);
     }
-
-    private void illegalMutation(T previousValue, T newValue) throws CoderException {
-      throw new IllegalMutationException(
-          String.format("Value %s mutated illegally, new value was %s."
-              + " Encoding was %s, now %s.",
-              previousValue, newValue,
-              CoderUtils.encodeToBase64(coder, previousValue),
-              CoderUtils.encodeToBase64(coder, newValue)),
-          previousValue, newValue);
+    private String encodeToBase64 (StructuralByteArray stream) {
+        return BaseEncoding.base64Url().omitPadding().encode(stream.getValue());
     }
+    private void illegalMutation(Object previousStructuralValue,
+                                   Object newStructuralValue) throws CoderException {
+      String previousValue;
+      String newValue;
+      if (previousStructuralValue instanceof StructuralByteArray) {
+          previousValue = encodeToBase64((StructuralByteArray) previousStructuralValue);
+          newValue = encodeToBase64((StructuralByteArray) newStructuralValue);
+        } else {
+          previousValue = CoderUtils.encodeToBase64(coder, (T) previousStructuralValue);
```

Comment:
There is no guarantee that this will work since the Coder is able to return any type, not necessarily a type which is encodable by the Coder. The original code kept the encoded version of this for that reason.